libcudacxx/include/cuda/std/array

```diff
@@ -22,7 +22,497 @@
+  {
+    _CCCL_ASSERT(__n < _Size, "out-of-bounds access in std::array<T, N>");
+    return __elems_[__n];
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr reference at(size_type __n)
+  {
+    if (__n >= _Size)
+    {
+      _CUDA_VSTD::__throw_out_of_range("array::at");
+    }
+    return __elems_[__n];
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const_reference at(size_type __n) const
+  {
+    if (__n >= _Size)
+    {
+      _CUDA_VSTD::__throw_out_of_range("array::at");
+    }
+    return __elems_[__n];
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr reference front() noexcept
+  {
+    return (*this)[0];
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const_reference front() const noexcept
+  {
+    return (*this)[0];
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr reference back() noexcept
+  {
+    return (*this)[_Size - 1];
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const_reference back() const noexcept
+  {
+    return (*this)[_Size - 1];
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr value_type* data() noexcept
+  {
+    return __elems_;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const value_type* data() const noexcept
+  {
+    return __elems_;
+  }
+};
+
+template <class _Tp>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT array<_Tp, 0>
+{
+  // types:
+  using __self                 = array;
+  using value_type             = _Tp;
+  using reference              = value_type&;
+  using const_reference        = const value_type&;
+  using iterator               = value_type*;
+  using const_iterator         = const value_type*;
+  using pointer                = value_type*;
+  using const_pointer          = const value_type*;
+  using size_type              = size_t;
+  using difference_type        = ptrdiff_t;
+  using reverse_iterator       = _CUDA_VSTD::reverse_iterator<iterator>;
+  using const_reverse_iterator = _CUDA_VSTD::reverse_iterator<const_iterator>;
+
+  using _CharType = conditional_t<_CCCL_TRAIT(is_const, _Tp), const char, char>;
+
+  struct _ArrayInStructT
+  {
+    _Tp __data_[1];
+  };
+  _CCCL_ALIGNAS_TYPE(_ArrayInStructT) _CharType __elems_[sizeof(_ArrayInStructT)];
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr value_type* data() noexcept
+  {
+    return nullptr;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const value_type* data() const noexcept
+  {
+    return nullptr;
+  }
+
+  // No explicit construct/copy/destroy for aggregate type
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr void fill(const value_type&)
+  {
+    static_assert(!_CCCL_TRAIT(is_const, _Tp), "cannot fill zero-sized array of type 'const T'");
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr void swap(array&) noexcept
+  {
+    static_assert(!_CCCL_TRAIT(is_const, _Tp), "cannot swap zero-sized array of type 'const T'");
+  }
+
+  // iterators:
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr iterator begin() noexcept
+  {
+    return iterator(nullptr);
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr const_iterator begin() const noexcept
```

Comment:
While we are at it moving the implementation, could we add [[nodiscard]] to all the pure accessors?